bookmarkbutton.py

```diff
@@ -0,0 +1,162 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Contact information:
+# Utkarsh Tiwari    iamutkarshtiwari@gmail.com
+
+"""
+BETA.
+"""
+
+import logging
+
+from gi.repository import GObject
+from gi.repository import Gtk
+
+from sugar3.graphics import style
+from sugar3.graphics.icon import Icon
+from sugar3.graphics.palette import Palette, ToolInvoker
+
+
+def _add_accelerator(tool_button):
+    if not tool_button.props.accelerator or not tool_button.get_toplevel() or \
+            not tool_button.get_child():
+        return
+
+    # TODO: should we remove the accelerator from the prev top level?
+    if not hasattr(tool_button.get_toplevel(), 'sugar_accel_group'):
+        logging.debug('No Gtk.AccelGroup in the top level window.')
+        return
+
+    accel_group = tool_button.get_toplevel().sugar_accel_group
+    keyval, mask = Gtk.accelerator_parse(tool_button.props.accelerator)
+    # the accelerator needs to be set at the child, so the Gtk.AccelLabel
+    # in the palette can pick it up.
+    accel_flags = Gtk.AccelFlags.LOCKED | Gtk.AccelFlags.VISIBLE
+    tool_button.get_child().add_accelerator('clicked', accel_group,
+                                            keyval, mask, accel_flags)
+
+
+def _hierarchy_changed_cb(tool_button, previous_toplevel):
+    _add_accelerator(tool_button)
+
+
+def setup_accelerator(tool_button):
+    _add_accelerator(tool_button)
+    tool_button.connect('hierarchy-changed', _hierarchy_changed_cb)
+
+
+class BookmarkToolButton(Gtk.ToggleToolButton):
+
+    __gtype_name__ = 'BookmarkToggleToolButton'
+
+    def __init__(self, icon_name=None, xo_color=None):
+        GObject.GObject.__init__(self)
+
+        self._palette_invoker = ToolInvoker(self)
+
+        if icon_name:
+            self.set_icon_name(icon_name)
+            self.icon_name = icon_name
+
+        self.connect('destroy', self.__destroy_cb)
+
+    def __destroy_cb(self, icon):
+        if self._palette_invoker is not None:
+            self._palette_invoker.detach()
+
+    def set_icon_name(self, icon_name):
+        icon = Icon(icon_name=icon_name)
+        self.set_icon_widget(icon)
+        icon.show()
+
+    def get_icon_name(self):
+        if self.props.icon_widget is not None:
+            return self.props.icon_widget.props.icon_name
+        else:
+            return None
+
+    icon_name = GObject.property(type=str, setter=set_icon_name,
+                                 getter=get_icon_name)
+
+    def set_xo_color(self, xo_color=None):
+        if xo_color is None:
+            icon = Icon(icon_name=self.icon_name)
+        else:
+            icon = Icon(icon_name=self.icon_name, xo_color=xo_color,
+                        pixel_size=style.STANDARD_ICON_SIZE)
```

Comment:
I'm amusing that this is the function that you added compared to the original toolbutton?  Maybe a better approach is to:

1.  Send a patch in to sugar-toolkit-gtk3 with this change (it is one of those, "Didn't we have this api already?", changes)
2.  Have a conditional import and include a minimum viable subclass with your changes